mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java

```diff
@@ -1176,19 +1178,23 @@ private Query createQuery(Criterion criterion, Query inMailboxes, Collection<Mes
     }
 
     @Override
-    public void add(MailboxSession session, Mailbox mailbox, MailboxMessage membership) throws IOException, MimeException {
-        Document doc = createMessageDocument(session, membership);
-        Document flagsDoc = createFlagsDocument(membership);
-
-        writer.addDocument(doc);
-        writer.addDocument(flagsDoc);
+    public Mono<Void> add(MailboxSession session, Mailbox mailbox, MailboxMessage membership) {
+        return Mono.fromRunnable(Throwing.runnable(() -> {
+            Document doc = createMessageDocument(session, membership);
+                Document flagsDoc = createFlagsDocument(membership);
+
+                writer.addDocument(doc);
+                writer.addDocument(flagsDoc);
+        }));
     }
 
     @Override
-    public void update(MailboxSession session, Mailbox mailbox, List<UpdatedFlags> updatedFlagsList) throws IOException {
-        for (UpdatedFlags updatedFlags : updatedFlagsList) {
-            update(mailbox, updatedFlags.getUid(), updatedFlags.getNewFlags());
-        }
+    public Mono<Void> update(MailboxSession session, Mailbox mailbox, List<UpdatedFlags> updatedFlagsList) {
+        return Mono.fromRunnable(Throwing.runnable(() -> {
+            for (UpdatedFlags updatedFlags : updatedFlagsList) {
```

Comment:
I think we can use more reactor ways that for loops?

Reply:
That would bring little value IMO as the underlying steps are fully synchronous.

Feel free to contribute it once this work is merged.